src/Orchard/Tasks/Locking/Services/DistributedLockService.cs

```diff
@@ -0,0 +1,153 @@
+using System;
+using System.Data;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Autofac;
+using Orchard.Data;
+using Orchard.Environment;
+using Orchard.Exceptions;
+using Orchard.Logging;
+using Orchard.Services;
+using Orchard.Tasks.Locking.Records;
+
+namespace Orchard.Tasks.Locking.Services {
+
+    public class DistributedLockService : IDistributedLockService {
+        private readonly IMachineNameProvider _machineNameProvider;
+        private readonly ILifetimeScope _lifetimeScope;
+        private readonly IClock _clock;
+        private readonly object _semaphore = new object();
+
+        public DistributedLockService(IMachineNameProvider machineNameProvider, ILifetimeScope lifetimeScope, IClock clock) {
+            _machineNameProvider = machineNameProvider;
+            _lifetimeScope = lifetimeScope;
+            _clock = clock;
+            Logger = NullLogger.Instance;
+        }
+
+        public ILogger Logger { get; set; }
+
+        public bool TryAcquireLock(string name, TimeSpan maxLifetime, TimeSpan timeout, out IDistributedLock @lock) {
+            lock (_semaphore) {
+                @lock = default(IDistributedLock);
+
+                try {
+                    var waitedTime = TimeSpan.Zero;
+                    var waitTime = TimeSpan.FromMilliseconds(timeout.TotalMilliseconds / 10);
```

Comment:
The idea is that if acquiring a lock fails, we want to keep retrying until a *total* of `timeout` time has elapsed. I took this code from someone else and reasoned it would work well, even for small numbers, since worst case the `Task.Delay` would be invoked right after the previous invocation, but only for a short amount of time.

Another function could work too I suppose, but I couldn't come up with anything else that is less arbitrary. I'm open to suggestions though.